### PR TITLE
Use ruff to replace black,  isort, pydocstyle & pycodestyle

### PR DESCRIPTION
**Pull request by marcospri**
A slight change of plans, as this generates a relatively big diff let's do it first and tackle linting afterwards. 

This change doesn't take much effort and reverting is also very easy, just switch the dependencies and run black/isort again.

https://docs.astral.sh/ruff/formatter/
https://docs.astral.sh/ruff/formatter/#intentional-deviations

### CODE REVIEW

**Comment by marcospri** [comment]
@seanh Should we port this to the cookiecutter right now or finish to replace pylint as well?

**Comment by seanh** [comment]
> @seanh Should we port this to the cookiecutter right now or finish to replace pylint as well?

Either way, I think we could port it right now if you like. The cookiecutter's script for automatically sending PRs to all our repos is currently broken: https://hypothes-is.slack.com/archives/C4K6M7P5E/p1714654897191519?thread_ts=1714654792.996779&cid=C4K6M7P5E (I've messaged GitHub support about it, no reply yet). But if this were added to the cookiecutter right now, we could just run `make template` in each repo one-by-one (whenever we feel like it for each repo) and get `make ruff`

File: pyproject.toml [hunk 1]
```diff
@@ -138,6 +130,11 @@ good-names = [
 output-format = "colorized"
 score = "no"
 
+[tool.ruff]
+target-version = "py311"
```
Comment (seanh) [inline pyproject.toml:134]: The Python version will have to be templated in the cookiecutter

**Review by seanh — APPROVED** [review]

**Comment by marcospri** [comment]
Included here also the change to replace pydocstyle and pycodestyle with ruff.

**Pull request merged**